spec/Methods.tex

```diff
@@ -0,0 +1,386 @@
+passed by reference, allowing modifications to \chpl{this}.  If
+\sntx{this-intent} is \chpl{const ref}, the receiver argument is passed
+by reference but it cannot be modified inside the method. The
+\sntx{this-intent} can also describe an abstract intent as follows. If it is
+\chpl{const}, the receiver argument will be passed with \chpl{const}
+intent. If it is left out entirely, the receiver will be passed with
+a default intent. For records, that default intent is \chpl{ref} if
+\chpl{this} is modified within the function and \chpl{const ref}
+otherwise.  For other types, the default \chpl{this} intent matches the
+default argument intent described in \rsec{The_Default_Intent}.
+
+\begin{chapelexample}{refThisIntent.chpl}
+In the following code, the \chpl{doubleMe} function is defined with a
+\sntx{this-intent} of \chpl{ref}, allowing variables of type \chpl{int} to
+double themselves.
+\begin{chapel}
+proc ref int.doubleMe() { this *= 2; }
+\end{chapel}
+\begin{chapelpost}
+var x: int = 2;
+x.doubleMe();
+writeln(x);
+\end{chapelpost}
+\begin{chapeloutput}
+4
+\end{chapeloutput}
+Given a variable \chpl{x = 2}, a call to \chpl{x.doubleMe()} will set \chpl{x}
+to \chpl{4}.
+\end{chapelexample}
+
+\section{The {\em this} Method}
+\label{The_this_Method}
+\index{methods!indexing}
+\index{this@\chpl{this}}
+\index{methods!this@\chpl{this}}
+
+A procedure method declared with the name \chpl{this} allows the receiver to be
+``indexed'' similarly to how an array is indexed.  Indexing (as with
+\chpl{A[1]} has the semantics of calling a method
+named \chpl{this}.  There is no other way to call a method
+called \chpl{this}.  The \chpl{this} method must be declared with
+parentheses even if the argument list is empty.
+
+\begin{chapelexample}{thisMethod.chpl}
+In the following code, the \chpl{this} method is used to create a
+class that acts like a simple array that contains three integers
+indexed by 1, 2, and 3.
+\begin{chapel}
+class ThreeArray {
+  var x1, x2, x3: int;
+  proc this(i: int) ref {
+    select i {
+      when 1 do return x1;
+      when 2 do return x2;
+      when 3 do return x3;
+    }
+    halt("ThreeArray index out of bounds: ", i);
+  }
+}
+\end{chapel}
+\begin{chapelpost}
+var ta = new ThreeArray();
+ta(1) = 1;
+ta(2) = 2;
+ta(3) = 3;
+for i in 1..3 do
+  writeln(ta(i));
+ta(4) = 4;
+\end{chapelpost}
+\begin{chapeloutput}
+1
+2
+3
+thisMethod.chpl:9: error: halt reached - ThreeArray index out of bounds: 4
+\end{chapeloutput}
+\end{chapelexample}
+
+\section{The {\em these} Method}
+\label{The_these_Method}
+\index{methods!iterating}
+\index{these@\chpl{these}}
+\index{methods!these@\chpl{these}}
+
+An iterator method declared with the name \chpl{these} allows the
+receiver to be
+``iterated over'' similarly to how a domain or array supports iteration.
+Using a value in the context of a loop where
+an \sntx{iteratable-expression} is expected has the semantics of calling
+a method on the value named \chpl{these}.
```

Comment:
The line breaks on this paragraph should probably be fixed, though it won't impact the generated pdf